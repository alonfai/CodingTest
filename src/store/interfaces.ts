import { Interfaces as API } from 'api';

/**
 * User input to retrieve city weather information
 */
export interface LocationInput {
  /**
   * City name
   */
  cityName: string;
  /**
   * ISO 3166 country code
   */
  countryCode: string;
}

/**
 * Exception data on API fetch request for city weather
 */
export interface ExceptionData extends LocationInput {
  /**
   * error message returned from the server
   */
  message: string;
}

/**
 * City condition interface data in UI
 */
export interface Condition extends Omit<API.Weather, 'icon'> {
  /**
   * URL for the icon of the weather condition
   */
  icon_url: string;
}

/**
 * Weather data for a city in the store
 */
export interface CityData {
  /**
   * Name of city
   */
  name: string;
  /**
   * Current temprature of the ciuty
   */
  temprature: number;
  /**
   * Weather condition
   */
  conditions: Condition[];
  /**
   * rain information
   */
  rain?: API.Rain;
  /**
   * snow information
   */
  snow?: API.Snow;
  /**
   * wind information
   */
  wind?: API.Wind;
}

/**
 * Full data object for a city with information, loading state and an optional error exceptions from API (if occured while fetching the data)
 */
export interface CityItem {
  /**
   * data collection
   */
  data?: CityData;
  /**
   * Does the
   */
  loading: boolean;

  /**
   * Exception data returned from the server
   */
  error?: string;
}

/**
 * City reducer data reprentation with key of "{city_name}_{country_code}" and value is actual api response
 */
export interface CityStore {
  [id: string]: CityItem;
}

/**
 * Reference to the store main data type
 */
export interface ReduxStore {
  /**
   * Reducer with all cities information
   */
  cities: CityStore;
}
