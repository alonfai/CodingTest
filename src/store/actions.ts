import { createAction } from 'redux-starter-kit';
import { Interfaces as API } from 'api';
import { LocationInput, ExceptionData } from './interfaces';

/**
 * Action that dispatch loading of city weather information
 */
export const LOAD_CITY_INFORMATION = createAction<LocationInput>('weather/get');

/**
 * SUCCESS response action when API return with city weather information
 */
export const LOAD_CITY_INFORMATION_SUCCESS = createAction<API.ApiResponse>('weather/get_success');

/**
 * FAILURE response action when API return with an error object on retriving city weather information
 */
export const LOAD_CITY_INFORMATION_FAILURE = createAction<ExceptionData>('weather/get_failure');
