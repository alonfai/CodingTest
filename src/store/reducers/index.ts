import city from './city';

export default {
  cities: city,
};
