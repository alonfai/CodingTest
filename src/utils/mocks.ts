import { Interfaces as API } from 'api';

export function getCoordsData(): API.GeoLocation {
  return {
    lat: 37.39,
    lon: -122.08,
  };
}

export function getConditions(): API.Weather[] {
  return [
    {
      id: 701,
      main: 'Mist',
      description: 'mist',
      icon: '50d',
    },
    {
      id: 300,
      main: 'Drizzle',
      description: 'light intensity drizzle',
      icon: '09d',
    },
  ];
}

export function getMainData(): API.Main {
  return {
    temp: 296.71,
    pressure: 1013,
    humidity: 53,
    temp_min: 294.82,
    temp_max: 298.71,
  };
}

export function getWindData(): API.Wind {
  return {
    speed: 1.5,
    deg: 350,
  };
}

export function getCloudData(): API.Cloud {
  return {
    all: 1,
  };
}

export function getRainData(): API.Rain {
  return {
    '1h': 10,
    '3h': 5,
  };
}

export function getSnowData(): API.Snow {
  return {
    '1h': 10,
    '3h': 25,
  };
}

export function getSysData(): API.System {
  return {
    type: 1,
    id: 5122,
    message: 0.0139,
    country: 'US',
    sunrise: 1560343627,
    sunset: 1560396563,
  };
}

/**
 * Get a mock version of the store with list of customers
 */
export function getWeatherInformation(): API.ApiResponse {
  return {
    base: 'stations',
    clouds: getCloudData(),
    coord: getCoordsData(),
    timezone: -25200,
    id: 420006353,
    name: 'Mountain View',
    cod: 200,
    dt: 1560350645,
    main: getMainData(),
    rain: getRainData(),
    snow: getSnowData(),
    sys: getSysData(),
    weather: getConditions(),
    wind: getWindData(),
  };
}
