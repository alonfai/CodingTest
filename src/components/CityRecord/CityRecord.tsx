import React from 'react';
import styled from 'styled-components';
import { Spinner, Alert, Image } from 'react-bootstrap';
import { v4 } from 'uuid';

import { Interfaces } from 'store';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
`;

const ListWrapper = styled.ul`
  font-size: 1rem;
`;

const ConditionsContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
`;

const TextWrapper = styled.li`
  font-size: 1.5rem;
`;

export interface ComponentProps {
  /**
   * City information item
   */
  item: Interfaces.CityItem;
}

export const CityRecord: React.FC<ComponentProps> = ({ item }) => {
  const { loading, data, error } = item;
  if (loading) {
    return (
      <Container data-testid="loading">
        <Spinner animation="border" role="status">
          <span className="sr-only">Loading...</span>
        </Spinner>
      </Container>
    );
  } else if (error || !data) {
    return (
      <Container data-testid="error">
        <Alert variant="danger">{error}</Alert>
      </Container>
    );
  } else {
    return (
      <Container data-testid="record">
        <ListWrapper>
          <TextWrapper>Temprature: {data.temprature} Â°</TextWrapper>
          {data.conditions.map(condition => {
            return (
              <ConditionsContainer key={v4()}>
                <TextWrapper>
                  Conditions: {condition.description} <Image src={condition.icon_url} roundedCircle />
                </TextWrapper>
              </ConditionsContainer>
            );
          })}
          {data.wind && (
            <TextWrapper>
              Wind: {data.wind.deg} meter/sec, Direction: {data.wind.deg} degrees
            </TextWrapper>
          )}
          {data.rain && (
            <TextWrapper>
              Rain: {data.rain['1h'] && `${data.rain["1h"]} mm for the last 1 hour,`} {data.rain['3h'] && `${data.rain["3h"]} mm for the last 3 hours`}
            </TextWrapper>
          )}
          {data.snow && (
            <TextWrapper>
              Snow: {data.snow['1h'] && `${data.snow["1h"]} mm for the last 1 hour,`} {data.snow['3h'] && `${data.snow["3h"]} mm for the last 3 hours`}
            </TextWrapper>
          )}
        </ListWrapper>
      </Container>
    );
  }
};

export default React.memo(CityRecord);
