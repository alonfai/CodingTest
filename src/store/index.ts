import { configureStore } from 'redux-starter-kit';
import * as Interfaces from './interfaces';
import * as actions from './actions';
import rootReducer from './reducers';

const store = configureStore<Interfaces.ReduxStore>({
  reducer: rootReducer,
});

// HMR Support
if (process.env.NODE_ENV === 'development' && (module as any).hot) {
  (module as any).hot.accept('./reducers', () => {
    const newRootReducer = require('./reducers').default;
    store.replaceReducer(newRootReducer);
  });
}

/**
 * Get the type of the store dispatch (different from regular useDispatch() hook interface, as inside the store we use "redux-thunk" as the async middleware)
 */
export type AppDispatch = typeof store.dispatch;

export { Interfaces, actions };

export default store;
