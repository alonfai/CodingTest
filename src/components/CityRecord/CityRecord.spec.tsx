import React from 'react';
import { render, RenderResult } from '@testing-library/react';

import * as Mocks from 'utils/mocks';
import { Interfaces } from 'store';
import { CityRecord, ComponentProps } from './CityRecord';

describe('CityRecord', () => {
  const defaultProps: ComponentProps = {
    item: {
      loading: false,
      error: undefined,
      data: {
        name: 'sydney',
        rain: Mocks.getRainData(),
        snow: Mocks.getSnowData(),
        temprature: Mocks.getMainData().temp,
        conditions: Mocks.getConditions().map(c => {
          const { icon } = c;
          delete c.icon;
          const condition: Interfaces.Condition = {
            ...c,
            icon_url: `http://www.icons.com/${icon}`,
          };
          return condition;
        }),
        wind: Mocks.getWindData(),
      },
    },
  };
  let renderer: RenderResult;

  const setup = (props = defaultProps) => {
    return render(<CityRecord {...props} />);
  };
  beforeEach(() => {
    renderer = setup();
  });

  it('should render successfully', () => {
    const { container } = renderer;
    expect(container.firstChild).toMatchSnapshot();
  });

  it('check when an error occurs, will show an alert error', () => {
    const props: ComponentProps = {
      ...defaultProps,
      item: {
        ...defaultProps.item,
        error: 'Error',
      },
    };
    renderer = setup(props);
    const { container } = renderer;
    expect(container).toBeTruthy();
    const alertItem = renderer.getByRole('alert');
    expect(alertItem).toBeTruthy();
  });

  it('check when data return is empty, will show an alert error', () => {
    const props: ComponentProps = {
      ...defaultProps,
      item: {
        ...defaultProps.item,
        error: 'Error',
        data: undefined,
      },
    };
    renderer = setup(props);
    const { container } = renderer;
    expect(container).toBeTruthy();
    const alertItem = renderer.getByRole('alert');
    expect(alertItem).toBeTruthy();
  });

  it('check when loading is set to true, show a spinner', () => {
    const props: ComponentProps = {
      ...defaultProps,
      item: {
        ...defaultProps.item,
        loading: true,
      },
    };
    renderer = setup(props);
    const { container } = renderer;
    expect(container).toBeTruthy();
    const loadingItem = renderer.getByText('Loading...');
    expect(loadingItem).toBeTruthy();
  });
});
