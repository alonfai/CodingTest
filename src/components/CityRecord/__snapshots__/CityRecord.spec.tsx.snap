// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityRecord should render successfully 1`] = `
<div
  class="sc-bdVaJa eUlSw"
  data-testid="record"
>
  <ul
    class="sc-bwzfXH fVRdXe"
  >
    <li
      class="sc-bxivhb kJBUIm"
    >
      Temprature: 
      296.71
       Â°
    </li>
    <div
      class="sc-htpNat dHjiyz"
    >
      <li
        class="sc-bxivhb kJBUIm"
      >
        Conditions: 
        mist
         
        <img
          class="rounded-circle"
          src="http://www.icons.com/50d"
        />
      </li>
    </div>
    <div
      class="sc-htpNat dHjiyz"
    >
      <li
        class="sc-bxivhb kJBUIm"
      >
        Conditions: 
        light intensity drizzle
         
        <img
          class="rounded-circle"
          src="http://www.icons.com/09d"
        />
      </li>
    </div>
    <li
      class="sc-bxivhb kJBUIm"
    >
      Wind: 
      350
       meter/sec, Direction: 
      350
       degrees
    </li>
    <li
      class="sc-bxivhb kJBUIm"
    >
      Rain: 
      10 mm for the last 1 hour,
       
      5 mm for the last 3 hours
    </li>
    <li
      class="sc-bxivhb kJBUIm"
    >
      Snow: 
      10 mm for the last 1 hour,
       
      25 mm for the last 3 hours
    </li>
  </ul>
</div>
`;
